dev/ProjectReunion_BootstrapDLL/MddBootstrap.cpp

```diff
@@ -0,0 +1,141 @@
+    RETURN_IF_WIN32_ERROR(OpenPackageInfoByFullName(packageFullName, 0, &packageInfoReference));
+    UINT32 bufferLength = 0;
+    const auto hr = HRESULT_FROM_WIN32(GetPackageInfo(packageInfoReference.get(), PACKAGE_FILTER_DIRECT, &bufferLength, nullptr, nullptr));
+    RETURN_HR_IF(hr, hr != HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER));
+    packageInfoBuffer = wil::make_unique_cotaskmem_nothrow<BYTE[]>(bufferLength);
+    RETURN_IF_NULL_ALLOC(packageInfoBuffer);
+    UINT32 packageInfoCount = 0;
+    RETURN_IF_WIN32_ERROR(GetPackageInfo(packageInfoReference.get(), PACKAGE_FILTER_DIRECT, &bufferLength, packageInfoBuffer.get(), &packageInfoCount));
+
+    const PACKAGE_INFO* packageInfo = reinterpret_cast<const PACKAGE_INFO*>(packageInfoBuffer.get());
+    for (size_t index = 0; index < packageInfoCount; ++index, ++packageInfo)
+    {
+        PCWSTR frameworkPackageFamilyName = L"Microsoft.ProjectReunionFramework_8wekyb3d8bbwe";
+        if (CompareStringOrdinal(packageInfo->packageFamilyName, -1, frameworkPackageFamilyName, -1, TRUE) == CSTR_EQUAL)
+        {
+            frameworkPackageInfo = packageInfo;
+            return S_OK;
+        }
+    }
+
+    // We didn't find the framework package as a dependency of the package
+    // providing the LifetimeManager.
+    //
+    // Verify the package providing the LifetimeManager declares
+    // <PackageDependency> on the Project Reunion Framework package.
+    RETURN_WIN32(APPMODEL_ERROR_PACKAGE_NOT_AVAILABLE);
+}
+
+HRESULT AddFrameworkToPath(PCWSTR frameworkPath)
+{
+    // Add the framework to the Loader's DllDirectory list
+    auto dllDirectoryCookie = AddDllDirectory(frameworkPath);
+    RETURN_LAST_ERROR_IF(dllDirectoryCookie == 0);
+
+    auto on_exit = wil::scope_exit([dllDirectoryCookie] {
+        if (dllDirectoryCookie)
+        {
+            RemoveDllDirectory(dllDirectoryCookie);
+        }
+    });
+
+    // Make a copy of the framework path to save for later
+    wil::unique_cotaskmem_string frameworkPathCopy(wil::make_cotaskmem_string_nothrow(frameworkPath));
+    RETURN_IF_NULL_ALLOC(frameworkPathCopy);
+
+    // Add the framework the the PATH environment variable
+    wil::unique_cotaskmem_string path;
+    RETURN_IF_FAILED(wil::GetEnvironmentVariableW(L"PATH", path));
+    if (path)
+    {
+        // PATH = frameworkPath + ";" + path
+        wil::unique_cotaskmem_string newPath;
+        RETURN_IF_FAILED(wil::str_concat_nothrow(newPath, frameworkPathCopy, L";", path));
+        RETURN_IF_WIN32_BOOL_FALSE(SetEnvironmentVariableW(L"PATH", newPath.get()));
+    }
+    else
+    {
+        const auto lastError = GetLastError();
+        RETURN_HR_IF(HRESULT_FROM_WIN32(lastError), lastError != ERROR_ENVVAR_NOT_FOUND);
+        RETURN_IF_WIN32_BOOL_FALSE(SetEnvironmentVariableW(L"PATH", frameworkPath));
+    }
+
+    g_frameworkPath = wistd::move(frameworkPathCopy);
+    g_dllDirectoryCookie = dllDirectoryCookie;
+    dllDirectoryCookie = 0;
```

Comment:
Use `on_exit.release()` to do this instead of a specific assignment.  Then you can remove the `if (dllDirectoryCookie)` check above.

Reply:
Solved by WIL-izing the cookie :-)